tests/test_cleanup.py

```diff
@@ -487,3 +507,810 @@ def test_file_vdi_delete(self, mock_lock):
 
         vdi.delete()
         mock_lock.Lock.cleanupAll.assert_called_with(str(vdi_uuid))
+
+    @mock.patch('cleanup.VDI', autospec=True)
+    @mock.patch('cleanup.SR._liveLeafCoalesce', autospec=True)
+    @mock.patch('cleanup.SR._snapshotCoalesce', autospec=True)
+    def test_coalesceLeaf(self, mock_srSnapshotCoalesce,
+                          mock_srLeafCoalesce, mock_vdi):
+
+        mock_vdi.canLiveCoalesce.return_value = True
+        mock_srLeafCoalesce.return_value = "This is a test"
+        sr_uuid = uuid4()
+        sr = create_cleanup_sr(uuid=str(sr_uuid))
+        vdi_uuid = uuid4()
+        vdi = cleanup.VDI(sr, str(vdi_uuid), False)
+
+        res = sr._coalesceLeaf(vdi)
+        self.assertEquals(res, "This is a test")
+        self.assertEqual(sr._liveLeafCoalesce.call_count, 1)
+        self.assertEqual(sr._snapshotCoalesce.call_count, 0)
+
+    @mock.patch('cleanup.VDI', autospec=True)
+    @mock.patch('cleanup.SR._liveLeafCoalesce', autospec=True)
+    @mock.patch('cleanup.SR._snapshotCoalesce', autospec=True)
+    def test_coalesceLeaf_coalesce_failed(self,
+                                          mock_srSnapshotCoalesce,
+                                          mock_srLeafCoalesce,
+                                          mock_vdi):
+
+        mock_vdi.canLiveCoalesce.return_value = False
+        mock_srSnapshotCoalesce.return_value = False
+        mock_srLeafCoalesce.return_value = False
+        sr_uuid = uuid4()
+        sr = create_cleanup_sr(uuid=str(sr_uuid))
+        vdi_uuid = uuid4()
+        vdi = cleanup.VDI(sr, str(vdi_uuid), False)
+
+        res = sr._coalesceLeaf(vdi)
+        self.assertEquals(res, False)
+
+    @mock.patch('cleanup.VDI.canLiveCoalesce',
+                autospec=True, return_value=False)
+    @mock.patch('cleanup.VDI.getSizeVHD',
+                autospec=True, return_value=1024)
+    @mock.patch('util.SMException.__init__',
+                autospec=True, return_value=None)
+    @mock.patch('cleanup.Util.log')
+    @mock.patch('cleanup.SR._snapshotCoalesce',
+                autospec=True, return_value=True)
+    def test_coalesceLeaf_size_the_same(self,
+                                        mock_srSnapshotCoalesce,
+                                        mock_log,
+                                        mock_exception_init,
+                                        mock_vdisize,
+                                        mockliveCoalesce):
+
+        sr_uuid = uuid4()
+        sr = create_cleanup_sr(uuid=str(sr_uuid))
+        vdi_uuid = uuid4()
+        vdi = cleanup.VDI(sr, str(vdi_uuid), False)
+
+        with self.assertRaises(util.SMException) as exc:
+            res = sr._coalesceLeaf(vdi)
+
+        mock_exception_init.assert_called_with("VDI {uuid} could"
```

Comment:
You should be able to test the content of `exc.message`, not sure if that's any easier.

Reply:
saves a mock, fixed.